Add REPL plugin hooks; Add `output`, `output-mode`, `stderr` attributes

JeffersGlass opened the pull request:
This PR does four major things

  - Add two new plugin lifecycle methods, `beforePyReplExec()` and `afterPyReplExec()`, which which are called immediately before and after a py-repl tag is evaluated, respectively.
  - Restore the `output` attribute of \<py-repl\> tags. If provided, the Element with the ID specified by output is used as:
    - The location to send writes to `sys.stdout`
    - The location where the result of evaluating the REPL (if any) is display()'d
  - Restore the `output-mode` attribute of \<py-repl\> tags. If the `output-mode == 'append'`, the contents of the output element (either specified by the `output` attribute or the default location as the next sibling of the REPL) are _not_ cleared before writing new contents. This is _not_ typical "REPL-like" behavior, but may be wanted if the user is already writing the REPL output somewhere else. _(Removed in [#881](https://github.com/pyscript/pyscript/pull/881).)_
  - Add a `stderr` attribute for \<py-repl\> tags. If provided, the Element with the ID specified by the output is used as the location to send writes to `sys.stderr`
  
All writes to the above locations (using `output` or `stderr` attributes) are _in addition_ to writing to the py-temrinal / ioMultiplexer.

---
This PR is a ready for review. The task list was:

 - [x] Add `beforePyReplExec()`, `afterPyReplExec()` to to plugins
 - [x] Expand StdioDirector plugin to accommodate new \<py-repl\> attributes
 - [x] Integration Tests:
   - [x] New integration tests for new attributes
   - [x] Update old plugin integration tests with new attributes
 - [x] Documentation
 - [x] Changelog Entries
 - [x] Review for completeness and correctness

  
  
Resolves #988, Fixes #1070, Resolves #975, Resolves #1074

JeffersGlass commented:
Apologies that this has stayed in draft for so long - I hope to get this completed and ready to review sometime this week.

JeffersGlass commented:
Something very odd is happening - I have three pending 'mergeable' checks that have failed but are still running, even after merging from main? Going to close this PR and open a fresh one.

FabioRosado commented on pyscriptjs/src/components/pyrepl.ts:
```diff
@@ -6,15 +6,17 @@ import { keymap } from '@codemirror/view';
-import { getAttribute, ensureUniqueId, htmlDecode } from '../utils';
-import type { Interpreter } from '../interpreter';
+import { getAttribute, ensureUniqueId, htmlDecode, createSingularWarning } from '../utils';
+import { Interpreter } from '../interpreter';
```
Should we still import this as type? 🤔

FabioRosado commented on pyscriptjs/src/main.ts:
```diff
@@ -192,9 +192,10 @@ export class PyScriptApp {
-        createCustomElements(interpreter);
 
-        initHandlers(interpreter);
+        //Takes a runtime and a reference to the PyScriptApp (to access plugins)
+        createCustomElements(runtime, this);
+        initHandlers(runtime);
```
Since we don't use `runtime` perhaps this needs to be updated?

```suggestion
        //Takes an interpreter and a reference to the PyScriptApp (to access plugins)
        createCustomElements(interpreter, this);
        initHandlers(interpreter);
```

JeffersGlass replied:
Thank you!

FabioRosado commented on pyscriptjs/src/plugin.ts:
```diff
@@ -60,6 +60,26 @@ export class Plugin {
+    /** The source of the <py-repl> tag has been fetched and its output-element determined;
+     * we're about to evaluate the source using the provided runtime
+     *
+     * @param options.interpreter The Runtime object that will be used to evaluated the Python source code
```
Do you think runtime here is okay or should we call it interpreter object? 🤔 

```suggestion
     * @param options.interpreter The Interpreter object that will be used to evaluate the Python source code
```

FabioRosado commented on pyscriptjs/src/plugin.ts:
```diff
@@ -60,6 +60,26 @@ export class Plugin {
+
+    /**
+     *
+     * @param options.interpreter  The Runtime object that will be used to evaluated the Python source code
```
Adding the suggestion if you think it should be interpreter 😄 
```suggestion
     * @param options.interpreter The Runtime object that will be used to evaluate the Python source code
```

FabioRosado commented on pyscriptjs/src/plugins/stdiodirector.ts:
```diff
@@ -53,4 +56,66 @@ export class StdioDirector extends Plugin {
+
+            }
+            else {
+                // 'otuput atribuite not provided
```
```suggestion
                // 'output attribute not provided
```

FabioRosado reviewed (approved):
Apologies for the LOOOOONG delay in reviewing this PR Jeff! Left some comments/suggestions and approving 😄

marimeireles reviewed (commented):
Just some minor comments here
🚀
Great work!

marimeireles commented:
Oh no, a lot changed 😕
@JeffersGlass lmk if you want I can take this over and rebase after I finish the event stuff.

JeffersGlass commented on pyscriptjs/src/plugins/stdiodirector.ts:
```diff
@@ -53,4 +56,66 @@ export class StdioDirector extends Plugin {
+            output_targeted_io = new TargetedStdio(options.pyReplTag, "output", true, true);
+        }
+        else {
+            output_targeted_io = new TargetedStdio(options.pyReplTag.outDiv, "id", true, true);
```
The PyRepl is responsible for setting up the outdiv - the correctedness of the outDiv is tested when the element is written to, not when then TargetedStdio is set up.

JeffersGlass commented:
@marimeireles @FabioRosado Finally got around to addressing the comments and merging from `main`! Would you mind doing a re-review?

For some reason, the Circle CI tests are stuck on a mergability check from a much earlier commit, and won't run the most recent tests at all. Any ideas how to clear those? Everything's merged and passing tests (locally) as of right now, for what it's worth. If they won't clear, I'll open a fresh PR with this branch again I guess.

marimeireles commented:
No idea why @JeffersGlass, sorry I missed your comment here.
To me it shows a bunch of unresolved conflicts.
Guess I'll give it a try on these :)

JeffersGlass commented:
> Got quite a few failing tests.

This is very odd to me - I merged from main yesterday, all tests passing locally, and pushed. I think the merge conflicts here are potentially an old circle ci artifact that won’t close? Going to try opening a fresh PR from the branch and see what happens…

marimeireles commented:
Hey @JeffersGlass, that would also work!
I've made a fresh PR here if you think it's easier and if it looks correct for you: https://github.com/pyscript/pyscript/pull/1301
But if the tests were passing locally to you then I might have merged something wrong... I was checking and it seems fine but will wait for you to come back to me so we don't dup. work.

marimeireles commented:
Ok great :)
This looks great!
I'll close my dupped PR and we continue from here =)

JeffersGlass commented:
Sounds good @marimeireles! I suspect I had messed up the inital merge/rebase, but this one is at least free of merge conflicts now as well.

marimeireles commented:
You can also see the local linting issues by running `pre-commit run --all-files`.

marimeireles reviewed (approved):
Went over it again and it looks good!
Thanks for coming back to this one @JeffersGlass, great work! 😊

marimeireles reviewed (changes requested):
Actually, there's one problem.. Let me check.

marimeireles reviewed (approved):
Okay, this is good now!

JeffersGlass commented:
Amazing! I'll get it merged later today!
